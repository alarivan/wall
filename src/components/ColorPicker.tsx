import React, { useState, useEffect, useRef } from 'react';
import { TColor } from '../types';
import { SketchPicker, ColorResult } from 'react-color';
import styled from 'styled-components';

export interface TriggerProps {
  style: React.CSSProperties;
  onClick: () => void;
}

interface Props {
  color?: TColor;
  onChangeComplete: (colorResult: ColorResult) => void;
  trigger: (props: TriggerProps) => JSX.Element;
}

const StyledDiv = styled.div`
  position: relative;
`;
const StyledPicker = styled(SketchPicker)`
  position: absolute;
  bottom: 0;
  right: 100%;
`;

const ColorPicker: React.FC<Props> = ({ color, onChangeComplete, trigger }) => {
  const [open, setOpen] = useState<boolean>(false);
  const wrapper = useRef<HTMLDivElement>(null);
  useEffect(() => {
    document.addEventListener('click', handleOutsideClick, false);

    return () => {
      document.removeEventListener('click', handleOutsideClick, false);
    };
  });

  const handleOutsideClick = (e: Event) => {
    const target = e.target as HTMLElement;
    if (wrapper.current && wrapper.current.contains(target)) return;

    setOpen(false);
  };

  const handleClick = () => setOpen(!open);

  return (
    <StyledDiv ref={wrapper}>
      {trigger({ style: { background: color }, onClick: handleClick })}
      {open && (
        <StyledPicker color={color} onChangeComplete={onChangeComplete} />
      )}
    </StyledDiv>
  );
};

export default ColorPicker;
