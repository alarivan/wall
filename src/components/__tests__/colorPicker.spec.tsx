import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import ColorPicker, { TriggerProps } from '../ColorPicker';

afterEach(cleanup);

describe('ColorPicker', () => {
  it('renders correctly', () => {
    const { asFragment } = render(
      <ColorPicker
        onChangeComplete={jest.fn()}
        trigger={jest.fn()}
        color='red'
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders trigger and calls onClick method on trigger', () => {
    const openColorPicker = jest.fn();
    const props = {
      onChangeComplete: jest.fn(),
      trigger: (props: TriggerProps) => (
        <button
          {...props}
          onClick={openColorPicker}
          data-testid='color-picker-trigger'
        ></button>
      ),
      color: 'red',
    };
    const { getByTestId } = render(<ColorPicker {...props} />);

    fireEvent.click(getByTestId('color-picker-trigger'));

    expect(openColorPicker).toHaveBeenCalledTimes(1);
  });
});
