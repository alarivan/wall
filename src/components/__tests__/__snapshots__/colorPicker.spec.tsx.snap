// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorPicker renders correctly 1`] = `
<DocumentFragment>
  .c0 {
  position: relative;
}

<div
    class="c0"
  />
</DocumentFragment>
`;
